src/main/java/org/medicmobile/webapp/mobile/PhotoGrabber.java

```diff
@@ -0,0 +1,203 @@
+	/**
+	 * This is actually the value of
+	 * {@code ImagePicker.DEFAULT_REQUEST_CODE}.  This value is not exposed
+	 * by the library, but needs to be supplied to
+	 * {@code getImageFromResult()} so that it properly returns the full-
+	 * size image.
+	 */
+	private static final int FAKE_REQUEST_CODE = 234;
+
+	private final Activity a;
+
+	private ValueCallback<Uri> uploadCallback;
+
+	PhotoGrabber(Activity a) {
+		this.a = a;
+	}
+
+//> EXTERNAL METHODS
+	boolean canHandle(String acceptType, boolean capture) {
+		return acceptType.startsWith("image/") && (!capture || canStartCamera());
+	}
+
+	void chooser(ValueCallback<Uri> callback, boolean capture) {
+		uploadCallback = callback;
+		if(capture) takePhoto();
+		else pickImage();
+	}
+
+	void process(int requestCode, int resultCode, Intent i) {
+		if(uploadCallback == null) {
+			warn(this, "uploadCallback is null for requestCode %s", requestCode);
+			return;
+		}
+
+		if(resultCode != RESULT_OK) {
+			// something went wrong; error code was logged above
+		} else {
+			try {
+				handleBitmapCallback(getImageFromResult(a, FAKE_REQUEST_CODE, resultCode, i));
+				return;
+			} catch(Exception ex) {
+				warn(ex, "process() :: error getting image from result");
+			}
+		}
+
+		handleNullCallback();
+	}
+
+//> PRIVATE HELPERS
+	private void handleNullCallback() {
+		uploadCallback.onReceiveValue(null);
+		uploadCallback = null;
+	}
+
+	private void handleBitmapCallback(final Bitmap bitmap) {
+		trace(this, "handleBitmapCallback() :: bitmap=%s", bitmap);
+		final ProgressDialog spinner = showSpinner(a, R.string.spnCompressingImage);
+		AsyncTask.execute(new Runnable() {
+			public void run() {
+				Uri uri = null;
+				try {
+					uri = writeToFile(bitmap);
+					trace(this, "process() :: image written to temp file.  uri=%s", uri);
+				} catch(Exception ex) {
+					warn(ex, "process() :: error writing image to temp file");
+				}
+				spinner.dismiss();
+
+				uploadCallback.onReceiveValue(uri);
+				uploadCallback = null;
+			}
+		});
+	}
+
+	private void takePhoto() {
+		a.startActivityForResult(cameraIntent(), GRAB_PHOTO);
+	}
+
+	private void pickImage() {
+		Intent i = getPickImageIntent(a, a.getString(R.string.promptChooseImage));
+		a.startActivityForResult(i, GRAB_PHOTO);
+	}
+
+	private boolean canStartCamera() {
+		return intentHandlerAvailableFor(a, cameraIntent());
+	}
+
+	private static Intent cameraIntent() {
+		return new Intent(ACTION_IMAGE_CAPTURE);
+	}
+
+	/**
+	 * Write the supplied {@code Bitmap} to a random file location.
+	 *
+	 * Android documentation suggests that while these are nominally "temp"
+	 * files, they may not be cleaned up automatically.  For now, this may
+	 * be considered a useful feature.
+	 *
+	 * @return the {@code android.net.Uri} of the created file
+	 */
+	private Uri writeToFile(Bitmap bitmap) throws IOException {
+		File temp = tempFile();
+		long tempFileSize = -1;
+		int quality;
+
+		do {
+			quality = 100;
```

Comment:
FWIW if this is JPEG quality levels I don't think you really ever want 100, it's kind of nuts for size vs quality[1]. I think 90 is a good starting quality, and almost certainly indistinguishable from 100.

[1] Stats from a random a 6000x4000 picture I have. 100 = 27.5mb, 90 = 9.2mb, 80 = 7mb, 70 = 5.2mb, etc.

Reply:
Is there any point going as low as `0`?  or should we run through `90`-`50` or something?

Reply:
Depends on the library, might be worth playing with. In any case I don't think 0 is actually valid (in imagemagick that means "default" I think, and then it guesses).

Unfortunately this stuff really depends on the situation that it's used in. I wouldn't store photos of my vacation with a jpeg quality of 30, but from how I understand it would be used, I think 30 is probably a good minimum.

For laughs, here is an image (scaled down to 8mpixels) with quality 1:

![q01](https://user-images.githubusercontent.com/583851/32267260-8e5808ae-bee2-11e7-9ff8-1fdaaf04d7b6.jpeg)

And here it is at 30:

![q30](https://user-images.githubusercontent.com/583851/32267271-97d06ba6-bee2-11e7-9164-0b8266b964bc.jpeg)

And here it is at 70:

![q70](https://user-images.githubusercontent.com/583851/32267296-adf65b66-bee2-11e7-8f9b-2306610c3f36.jpeg)

The image is kind of blurry, but then if we're getting into pixel peeping to see differences then I think our interest in quality has progressed further than our use case :-)

Reply:
Interesting comparison!  I've set the quality range from 90 down to 50, and after that the dimensions will both be halved, and quality set back to 90.  I'll change minimum quality to 30.